File: android-kit-base/src/main/java/com/mparticle/kits/mappings/EventWrapper.java
package com.mparticle.kits.mappings;

import com.mparticle.MPEvent;
import com.mparticle.commerce.CommerceEvent;
import com.mparticle.commerce.Product;
import com.mparticle.commerce.Promotion;
import com.mparticle.kits.CommerceEventUtils;
import com.mparticle.kits.KitUtils;

import java.util.AbstractMap;
import java.util.HashMap;
import java.util.Map;

/**
 * Decorator classes for MPEvent and CommerceEvent. Used to extend functionality and to cache values for Projection processing.
 */
abstract class EventWrapper {
    public abstract Map<Integer, String> getAttributeHashes();

    protected Map<Integer, String> attributeHashes;

    public abstract int getEventTypeOrdinal();

    public abstract Object getEvent();

    public abstract int getMessageType();

    public abstract int getEventHash();

    protected static Map<Integer, String> getHashes(String hashPrefix, Map<String, String> map) {
        Map<Integer, String> hashedMap = new HashMap<Integer, String>();
        for (Map.Entry<String, String> entry : map.entrySet()) {
            int hash = KitUtils.hashForFiltering(hashPrefix + entry.getKey());
            hashedMap.put(hash, entry.getKey());
        }
        return hashedMap;
    }

    public abstract Map.Entry<String, String> findAttribute(String propertyType, int hash, Product product, Promotion promotion);

    public abstract Map.Entry<String, String> findAttribute(String propertyType, String keyName, Product product, Promotion promotion);

    static class CommerceEventWrapper extends EventWrapper {
        private CommerceEvent mCommerceEvent;
        private Map<Integer, String> eventFieldHashes;
        private HashMap<String, String> eventFieldAttributes;

        public CommerceEventWrapper(CommerceEvent event) {
            this.mCommerceEvent = event;
        }

        @Override
        public Map<Integer, String> getAttributeHashes() {
            if (attributeHashes == null) {
                attributeHashes = new HashMap<Integer, String>();
                if (mCommerceEvent.getCustomAttributeStrings() != null) {
                    for (Map.Entry<String, String> entry : mCommerceEvent.getCustomAttributeStrings().entrySet()) {
                        int hash = KitUtils.hashForFiltering(getEventTypeOrdinal() + entry.getKey());
                        attributeHashes.put(hash, entry.getKey());
                    }
                }
            }
            return attributeHashes;
        }

        public int getEventTypeOrdinal() {
            return CommerceEventUtils.getEventType(mCommerceEvent);
        }

        public CommerceEvent getEvent() {
            return mCommerceEvent;
        }

        public void setEvent(CommerceEvent event) {
            mCommerceEvent = event;
        }

        public int getMessageType() {
            return 16;
        }

        public int getEventHash() {
            return KitUtils.hashForFiltering("" + getEventTypeOrdinal());
        }

        public Map.Entry<String, String> findAttribute(String propertyType, int hash, Product product, Promotion promotion) {
            if (CustomMapping.PROPERTY_LOCATION_EVENT_ATTRIBUTE.equalsIgnoreCase(propertyType)) {
                if (getEvent().getCustomAttributeStrings() == null || getEvent().getCustomAttributeStrings().size() == 0) {
                    return null;
                }
                String key = getAttributeHashes().get(hash);
                if (key != null) {
                    return new AbstractMap.SimpleEntry<String, String>(key, mCommerceEvent.getCustomAttributeStrings().get(key));
                }
            } else if (CustomMapping.PROPERTY_LOCATION_EVENT_FIELD.equalsIgnoreCase(propertyType)) {
                if (eventFieldHashes == null) {
                    if (eventFieldAttributes == null) {
                        eventFieldAttributes = new HashMap<String, String>();
                        CommerceEventUtils.extractActionAttributes(getEvent(), eventFieldAttributes);
                        CommerceEventUtils.extractTransactionAttributes(getEvent(), eventFieldAttributes);
                    }
                    eventFieldHashes = getHashes(getEventTypeOrdinal() + "", eventFieldAttributes);
                }
                String key = eventFieldHashes.get(hash);
                if (key != null) {
                    return new AbstractMap.SimpleEntry<String, String>(key, eventFieldAttributes.get(key));
                }
            } else if (CustomMapping.PROPERTY_LOCATION_PRODUCT_ATTRIBUTE.equalsIgnoreCase(propertyType)) {
                if (product == null || product.getCustomAttributes() == null || product.getCustomAttributes().size() == 0) {
                    return null;
                }
                Map<String, String> attributes = new HashMap<String, String>();
                CommerceEventUtils.extractProductAttributes(product, attributes);
                Map<Integer, String> hashes = getHashes(getEventTypeOrdinal() + "", attributes);
                String key = hashes.get(hash);
                if (key != null) {
                    return new AbstractMap.SimpleEntry<String, String>(key, attributes.get(key));
                }
            } else if (CustomMapping.PROPERTY_LOCATION_PRODUCT_FIELD.equalsIgnoreCase(propertyType)) {
                if (product == null) {
                    return null;
                }
                Map<String, String> attributes = new HashMap<String, String>();
                CommerceEventUtils.extractProductFields(product, attributes);
                Map<Integer, String> hashes = getHashes(getEventTypeOrdinal() + "", attributes);
                String key = hashes.get(hash);
                if (key != null) {
                    return new AbstractMap.SimpleEntry<String, String>(key, attributes.get(key));
                }
            } else if (CustomMapping.PROPERTY_LOCATION_PROMOTION_FIELD.equalsIgnoreCase(propertyType)) {
                if (promotion == null) {
                    return null;
                }
                Map<String, String> attributes = new HashMap<String, String>();
                CommerceEventUtils.extractPromotionAttributes(promotion, attributes);
                Map<Integer, String> hashes = getHashes(getEventTypeOrdinal() + "", attributes);
                String key = hashes.get(hash);
                if (key != null) {
                    return new AbstractMap.SimpleEntry<String, String>(key, attributes.get(key));
                }
            }
            return null;
        }

        @Override
        public Map.Entry<String, String> findAttribute(String propertyType, String keyName, Product product, Promotion promotion) {
            if (CustomMapping.PROPERTY_LOCATION_EVENT_ATTRIBUTE.equalsIgnoreCase(propertyType)) {
                if (getEvent().getCustomAttributeStrings() == null || getEvent().getCustomAttributeStrings().size() == 0) {
                    return null;
                }
                if (getEvent().getCustomAttributeStrings().containsKey(keyName)) {
                    return new AbstractMap.SimpleEntry<String, String>(keyName, getEvent().getCustomAttributeStrings().get(keyName));
                }
            } else if (CustomMapping.PROPERTY_LOCATION_EVENT_FIELD.equalsIgnoreCase(propertyType)) {
                if (eventFieldAttributes == null) {
                    eventFieldAttributes = new HashMap<String, String>();
                    CommerceEventUtils.extractActionAttributes(getEvent(), eventFieldAttributes);
                    CommerceEventUtils.extractTransactionAttributes(getEvent(), eventFieldAttributes);
                }
                if (eventFieldAttributes.containsKey(keyName)) {
                    return new AbstractMap.SimpleEntry<String, String>(keyName, eventFieldAttributes.get(keyName));
                }
            } else if (CustomMapping.PROPERTY_LOCATION_PRODUCT_ATTRIBUTE.equalsIgnoreCase(propertyType)) {
                if (product == null || product.getCustomAttributes() == null) {
                    return null;
                }
                Map<String, String> attributes = new HashMap<String, String>();
                CommerceEventUtils.extractProductAttributes(product, attributes);

                if (attributes.containsKey(keyName)) {
                    return new AbstractMap.SimpleEntry<String, String>(keyName, attributes.get(keyName));
                }
            } else if (CustomMapping.PROPERTY_LOCATION_PRODUCT_FIELD.equalsIgnoreCase(propertyType)) {
                if (product == null) {
                    return null;
                }
                Map<String, String> attributes = new HashMap<String, String>();
                CommerceEventUtils.extractProductFields(product, attributes);

                if (attributes.containsKey(keyName)) {
                    return new AbstractMap.SimpleEntry<String, String>(keyName, attributes.get(keyName));
                }
            } else if (CustomMapping.PROPERTY_LOCATION_PROMOTION_FIELD.equalsIgnoreCase(propertyType)) {
                if (promotion == null) {
                    return null;
                }
                Map<String, String> attributes = new HashMap<String, String>();
                CommerceEventUtils.extractPromotionAttributes(promotion, attributes);
                if (attributes.containsKey(keyName)) {
                    return new AbstractMap.SimpleEntry<String, String>(keyName, attributes.get(keyName));
                }
            }
            return null;
        }
    }

    static class MPEventWrapper extends EventWrapper {
        private final MPEvent mEvent;
        private boolean mScreenEvent;

        public MPEventWrapper(MPEvent event) {
            this(event, false);
        }

        public MPEventWrapper(MPEvent event, boolean isScreenEvent) {
            this.mEvent = event;
            this.mScreenEvent = isScreenEvent;
        }

        public Map<Integer, String> getAttributeHashes() {
            if (attributeHashes == null) {
                attributeHashes = new HashMap<Integer, String>();
                if (mEvent.getCustomAttributeStrings() != null) {
                    for (Map.Entry<String, String> entry : mEvent.getCustomAttributeStrings().entrySet()) {
                        int hash = KitUtils.hashForFiltering(getEventTypeOrdinal() + mEvent.getEventName() + entry.getKey());
                        attributeHashes.put(hash, entry.getKey());
                    }
                }

            }
            return attributeHashes;
        }

        public MPEvent getEvent() {
            return mEvent;
        }

        public int getEventTypeOrdinal() {
            if (mScreenEvent) {
                return 0;
            } else {
                return mEvent.getEventType().ordinal();
            }
        }

        public int getEventHash() {
            if (mScreenEvent) {
                return KitUtils.hashForFiltering(getEventTypeOrdinal() + mEvent.getEventName());
            } else {
                return mEvent.getEventHash();
            }
        }

        public int getMessageType() {
            if (mScreenEvent) {
                return 3;
            } else {
                return 4;
            }
        }

        public Map.Entry<String, String> findAttribute(String propertyType, String keyName, Product product, Promotion promotion) {
            if (CustomMapping.PROPERTY_LOCATION_EVENT_ATTRIBUTE.equalsIgnoreCase(propertyType)) {
                if (getEvent().getCustomAttributeStrings() == null) {
                    return null;
                }
                String value = getEvent().getCustomAttributeStrings().get(keyName);
                if (value != null) {
                    return new AbstractMap.SimpleEntry<String, String>(keyName, value);
                }
            }
            return null;
        }

        public Map.Entry<String, String> findAttribute(String propertyType, int hash, Product product, Promotion promotion) {
            if (CustomMapping.PROPERTY_LOCATION_EVENT_ATTRIBUTE.equalsIgnoreCase(propertyType)) {
                String key = getAttributeHashes().get(hash);
                if (key != null) {
                    return new AbstractMap.SimpleEntry<>(key, mEvent.getCustomAttributeStrings().get(key));
                }
            }
            return null;
        }

    }
}

File: android-kit-base/src/main/java/com/mparticle/kits/mappings/EventWrapper.kt
package com.mparticle.kits.mappings

import com.mparticle.MPEvent
import com.mparticle.commerce.CommerceEvent
import com.mparticle.commerce.Product
import com.mparticle.commerce.Promotion
import com.mparticle.kits.CommerceEventUtils
import com.mparticle.kits.KitUtils.hashForFiltering
import java.util.AbstractMap
import java.util.HashMap

/**
 * Decorator classes for MPEvent and CommerceEvent. Used to extend functionality and to cache values for Projection processing.
 */
abstract class EventWrapper {
    abstract fun getAttributeHashes(): Map<Int, String>
    protected var _attributeHashes: MutableMap<Int, String>? = null
    abstract val eventTypeOrdinal: Int
    abstract val event: Any?
    abstract val messageType: Int
    abstract val eventHash: Int
    abstract fun findAttribute(
        propertyType: String?,
        hash: Int,
        product: Product?,
        promotion: Promotion?
    ): Map.Entry<String, String?>?

    abstract fun findAttribute(
        propertyType: String?,
        keyName: String,
        product: Product?,
        promotion: Promotion?
    ): Map.Entry<String, String?>?

    class CommerceEventWrapper(private var mCommerceEvent: CommerceEvent) :
        EventWrapper() {
        private var eventFieldHashes: Map<Int, String>? = null
        private var eventFieldAttributes: HashMap<String, String?>? = null
        override fun getAttributeHashes(): Map<Int, String> {
            if (_attributeHashes == null) {
                _attributeHashes = HashMap()
                if (mCommerceEvent.customAttributeStrings != null) {
                    for ((key) in mCommerceEvent.customAttributeStrings!!) {
                        val hash = hashForFiltering(eventTypeOrdinal.toString() + key)
                        _attributeHashes!![hash] = key
                    }
                }
            }
            return _attributeHashes!!
        }

        override val eventTypeOrdinal: Int
            get() = CommerceEventUtils.getEventType(mCommerceEvent)
        override val event: CommerceEvent
            get() = mCommerceEvent

        fun setEvent(event: CommerceEvent) {
            mCommerceEvent = event
        }

        override val messageType: Int
            get() = 16
        override val eventHash: Int
            get() = hashForFiltering("" + eventTypeOrdinal)

        override fun findAttribute(
            propertyType: String?,
            hash: Int,
            product: Product?,
            promotion: Promotion?
        ): Map.Entry<String, String?>? {
            if (CustomMapping.Companion.PROPERTY_LOCATION_EVENT_ATTRIBUTE.equals(
                    propertyType,
                    ignoreCase = true
                )
            ) {
                if (mCommerceEvent.customAttributeStrings == null || mCommerceEvent.customAttributeStrings!!.size == 0) {
                    return null
                }
                val key = getAttributeHashes()[hash]
                if (key != null) {
                    return AbstractMap.SimpleEntry(
                        key,
                        mCommerceEvent.customAttributeStrings!![key]
                    )
                }
            } else if (CustomMapping.Companion.PROPERTY_LOCATION_EVENT_FIELD.equals(
                    propertyType,
                    ignoreCase = true
                )
            ) {
                if (eventFieldHashes == null) {
                    if (eventFieldAttributes == null) {
                        eventFieldAttributes = HashMap()
                        CommerceEventUtils.extractActionAttributes(
                            mCommerceEvent,
                            eventFieldAttributes!!
                        )
                        CommerceEventUtils.extractTransactionAttributes(
                            mCommerceEvent,
                            eventFieldAttributes!!
                        )
                    }
                    eventFieldHashes =
                        getHashes(eventTypeOrdinal.toString() + "", eventFieldAttributes!!)
                }
                val key = eventFieldHashes!![hash]
                if (key != null) {
                    return AbstractMap.SimpleEntry(key, eventFieldAttributes!![key])
                }
            } else if (CustomMapping.Companion.PROPERTY_LOCATION_PRODUCT_ATTRIBUTE.equals(
                    propertyType,
                    ignoreCase = true
                )
            ) {
                if (product == null || product.customAttributes == null || product.customAttributes!!.size == 0) {
                    return null
                }
                val attributes: MutableMap<String, String?> = HashMap()
                CommerceEventUtils.extractProductAttributes(product, attributes)
                val hashes = getHashes(
                    eventTypeOrdinal.toString() + "", attributes
                )
                val key = hashes[hash]
                if (key != null) {
                    return AbstractMap.SimpleEntry(key, attributes[key])
                }
            } else if (CustomMapping.Companion.PROPERTY_LOCATION_PRODUCT_FIELD.equals(
                    propertyType,
                    ignoreCase = true
                )
            ) {
                if (product == null) {
                    return null
                }
                val attributes: MutableMap<String, String?> = HashMap()
                CommerceEventUtils.extractProductFields(product, attributes)
                val hashes = getHashes(
                    eventTypeOrdinal.toString() + "", attributes
                )
                val key = hashes[hash]
                if (key != null) {
                    return AbstractMap.SimpleEntry(key, attributes[key])
                }
            } else if (CustomMapping.Companion.PROPERTY_LOCATION_PROMOTION_FIELD.equals(
                    propertyType,
                    ignoreCase = true
                )
            ) {
                if (promotion == null) {
                    return null
                }
                val attributes: MutableMap<String, String?> = HashMap()
                CommerceEventUtils.extractPromotionAttributes(promotion, attributes)
                val hashes = getHashes(
                    eventTypeOrdinal.toString() + "", attributes
                )
                val key = hashes[hash]
                if (key != null) {
                    return AbstractMap.SimpleEntry(key, attributes[key])
                }
            }
            return null
        }

        override fun findAttribute(
            propertyType: String?,
            keyName: String,
            product: Product?,
            promotion: Promotion?
        ): Map.Entry<String, String?>? {
            if (CustomMapping.Companion.PROPERTY_LOCATION_EVENT_ATTRIBUTE.equals(
                    propertyType,
                    ignoreCase = true
                )
            ) {
                if (mCommerceEvent.customAttributeStrings == null || mCommerceEvent.customAttributeStrings!!.size == 0) {
                    return null
                }
                if (mCommerceEvent.customAttributeStrings!!.containsKey(keyName)) {
                    return AbstractMap.SimpleEntry(
                        keyName,
                        mCommerceEvent.customAttributeStrings!![keyName]
                    )
                }
            } else if (CustomMapping.Companion.PROPERTY_LOCATION_EVENT_FIELD.equals(
                    propertyType,
                    ignoreCase = true
                )
            ) {
                if (eventFieldAttributes == null) {
                    eventFieldAttributes = HashMap()
                    CommerceEventUtils.extractActionAttributes(
                        mCommerceEvent,
                        eventFieldAttributes!!
                    )
                    CommerceEventUtils.extractTransactionAttributes(
                        mCommerceEvent,
                        eventFieldAttributes!!
                    )
                }
                if (eventFieldAttributes!!.containsKey(keyName)) {
                    return AbstractMap.SimpleEntry(keyName, eventFieldAttributes!![keyName])
                }
            } else if (CustomMapping.Companion.PROPERTY_LOCATION_PRODUCT_ATTRIBUTE.equals(
                    propertyType,
                    ignoreCase = true
                )
            ) {
                if (product == null || product.customAttributes == null) {
                    return null
                }
                val attributes: MutableMap<String, String?> = HashMap()
                CommerceEventUtils.extractProductAttributes(product, attributes)
                if (attributes.containsKey(keyName)) {
                    return AbstractMap.SimpleEntry(keyName, attributes[keyName])
                }
            } else if (CustomMapping.Companion.PROPERTY_LOCATION_PRODUCT_FIELD.equals(
                    propertyType,
                    ignoreCase = true
                )
            ) {
                if (product == null) {
                    return null
                }
                val attributes: MutableMap<String, String?> = HashMap()
                CommerceEventUtils.extractProductFields(product, attributes)
                if (attributes.containsKey(keyName)) {
                    return AbstractMap.SimpleEntry(keyName, attributes[keyName])
                }
            } else if (CustomMapping.Companion.PROPERTY_LOCATION_PROMOTION_FIELD.equals(
                    propertyType,
                    ignoreCase = true
                )
            ) {
                if (promotion == null) {
                    return null
                }
                val attributes: MutableMap<String, String?> = HashMap()
                CommerceEventUtils.extractPromotionAttributes(promotion, attributes)
                if (attributes.containsKey(keyName)) {
                    return AbstractMap.SimpleEntry(keyName, attributes[keyName])
                }
            }
            return null
        }
    }

    class MPEventWrapper @JvmOverloads constructor(
        private val mEvent: MPEvent,
        private val mScreenEvent: Boolean = false
    ) : EventWrapper() {
        override fun getAttributeHashes(): Map<Int, String> {
            if (_attributeHashes == null) {
                _attributeHashes = HashMap()
                if (mEvent.customAttributeStrings != null) {
                    for ((key) in mEvent.customAttributeStrings!!) {
                        val hash =
                            hashForFiltering(eventTypeOrdinal.toString() + mEvent.eventName + key)
                        _attributeHashes!![hash] = key
                    }
                }
            }
            return _attributeHashes!!
        }

        override val event: MPEvent
            get() = mEvent
        override val eventTypeOrdinal: Int
            get() = if (mScreenEvent) {
                0
            } else {
                mEvent.eventType.ordinal
            }
        override val eventHash: Int
            get() = if (mScreenEvent) {
                hashForFiltering(eventTypeOrdinal.toString() + mEvent.eventName)
            } else {
                mEvent.eventHash
            }
        override val messageType: Int
            get() = if (mScreenEvent) {
                3
            } else {
                4
            }

        override fun findAttribute(
            propertyType: String?,
            keyName: String,
            product: Product?,
            promotion: Promotion?
        ): Map.Entry<String, String?>? {
            if (CustomMapping.Companion.PROPERTY_LOCATION_EVENT_ATTRIBUTE.equals(
                    propertyType,
                    ignoreCase = true
                )
            ) {
                if (mEvent.customAttributeStrings == null) {
                    return null
                }
                val value = mEvent.customAttributeStrings!![keyName]
                if (value != null) {
                    return AbstractMap.SimpleEntry(keyName, value)
                }
            }
            return null
        }

        override fun findAttribute(
            propertyType: String?,
            hash: Int,
            product: Product?,
            promotion: Promotion?
        ): Map.Entry<String, String?>? {
            if (CustomMapping.Companion.PROPERTY_LOCATION_EVENT_ATTRIBUTE.equals(
                    propertyType,
                    ignoreCase = true
                )
            ) {
                val key = getAttributeHashes()[hash]
                if (key != null) {
                    return AbstractMap.SimpleEntry(key, mEvent.customAttributeStrings!![key])
                }
            }
            return null
        }
    }

    companion object {
        protected fun getHashes(hashPrefix: String, map: Map<String, String?>): Map<Int, String> {
            val hashedMap: MutableMap<Int, String> = HashMap()
            for ((key) in map) {
                val hash = hashForFiltering(hashPrefix + key)
                hashedMap[hash] = key
            }
            return hashedMap
        }
    }
}
